File: apps/angular-app/src/app/features/funds/data/repositories/fund.repository-impl.ts
import { Injectable, inject } from '@angular/core';
import { Observable, map, catchError, of, switchMap } from 'rxjs';
import { Result, right, left } from '../../../../core/errors/either';
import { ServerFailure, NetworkFailure } from '../../../../core/errors/failures';
import { Fund } from '../../domain/entities/fund.entity';
import { FundRepository } from '../../domain/repositories/fund.repository';
import { FundRemoteDatasource } from '../datasources/fund-remote.datasource';
import { FundMapper } from '../mappers/fund.mapper';

@Injectable({ providedIn: 'root' })
export class FundRepositoryImpl extends FundRepository {
  private readonly datasource = inject(FundRemoteDatasource);

  getFunds(): Observable<Result<Fund[]>> {
    return this.datasource.getFunds().pipe(
      map((dtos) => right<Fund[]>(FundMapper.fromDtoList(dtos))),
      catchError((error) => {
        if (error.status === 0) {
          return of(left<NetworkFailure>(new NetworkFailure()));
        }
        return of(left<ServerFailure>(new ServerFailure(error.message)));
      })
    );
  }

  subscribeTo(fundId: number, amount: number, notification: string): Observable<Result<void>> {
    return this.datasource.getUser().pipe(
      switchMap((user) => {
        const newBalance = user.balance - amount;
        const newSubscribed = [...user.subscribedFunds, fundId];
        return this.datasource.subscribe(fundId, amount, notification).pipe(
          switchMap(() => this.datasource.updateUser(newBalance, newSubscribed)),
          map(() => right<void>(undefined)),
        );
      }),
      catchError((error) => {
        return of(left<ServerFailure>(new ServerFailure(error.message)));
      })
    );
  }
}

File: apps/flutter-app/lib/features/funds/data/repositories/fund_repository_impl.dart
import 'package:dartz/dartz.dart';
import 'package:fondos_btg/core/error/exceptions.dart';
import 'package:fondos_btg/core/error/failures.dart';
import 'package:fondos_btg/features/funds/data/datasources/fund_remote_datasource.dart';
import 'package:fondos_btg/features/funds/domain/entities/fund.dart';
import 'package:fondos_btg/features/funds/domain/repositories/fund_repository.dart';

class FundRepositoryImpl implements FundRepository {
  final FundRemoteDatasource _remoteDatasource;

  const FundRepositoryImpl(this._remoteDatasource);

  @override
  Future<Either<Failure, List<Fund>>> getFunds() async {
    try {
      final dtos = await _remoteDatasource.getFunds();
      final funds = dtos.map((dto) => dto.toEntity()).toList();
      return Right(funds);
    } on ServerException catch (e) {
      return Left(ServerFailure(e.message));
    } on NotFoundException catch (e) {
      return Left(NotFoundFailure(e.message));
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }

  @override
  Future<Either<Failure, double>> subscribeToFund({
    required int fundId,
    required double amount,
    required String notificationMethod,
  }) async {
    try {
      final newBalance = await _remoteDatasource.subscribeToFund(
        fundId: fundId,
        amount: amount,
        notificationMethod: notificationMethod,
      );
      return Right(newBalance);
    } on BadRequestException catch (e) {
      if (e.message.contains('already subscribed') ||
          e.message.contains('ya vinculado')) {
        return Left(AlreadySubscribedFailure(e.message));
      }
      if (e.message.contains('insufficient') ||
          e.message.contains('saldo')) {
        return Left(InsufficientBalanceFailure(e.message));
      }
      return Left(ServerFailure(e.message));
    } on ServerException catch (e) {
      return Left(ServerFailure(e.message));
    } on NotFoundException catch (e) {
      return Left(NotFoundFailure(e.message));
    } catch (e) {
      return Left(ServerFailure(e.toString()));
    }
  }
}
